### PR TITLE
add verified_fields attribute to submission model

### PR DESCRIPTION
**Pull request by dodumosu**
this commit adds per-field verification on the database model, the QA queries and filters and inline QA computation (for displaying the result of a logical check in the QA list and submission edit views).

### CODE REVIEW

**Comment by takinbo** [comment]
rebase

File: apollo/submissions/forms.py [hunk 1]
```diff
@@ -125,15 +125,20 @@ def make_submission_edit_form_class(event, form):
...
-        form.form_type == 'CHECKLIST' and
-        permissions.edit_submission_quarantine_status.can()
-    ):
-        form_fields['quarantine_status'] = fields.SelectField(
-            choices=models.Submission.QUARANTINE_STATUSES,
-            filters=[lambda data: data if data else ''],
-            validators=[validators.Optional()]
-        )
+    if form.form_type == 'CHECKLIST':
+        if permissions.edit_submission_quarantine_status.can:
```
Comment (takinbo) [inline apollo/submissions/forms.py]: should be `permissions.edit_submission_quarantine_status.can()`?

File: apollo/submissions/forms.py [hunk 2]
```diff
@@ -125,15 +125,20 @@ def make_submission_edit_form_class(event, form):
...
+            form_fields['quarantine_status'] = fields.SelectField(
+                choices=models.Submission.QUARANTINE_STATUSES,
+                filters=[lambda data: data if data else ''],
+                validators=[validators.Optional()]
+            )
+        if permissions.edit_submission_quarantine_status.can():
+            form_fields['verified_fields'] = fields.SelectMultipleField(
+                'Verified fields',
```
Comment (takinbo) [inline apollo/submissions/forms.py]: Can we use Title Case instead?

File: apollo/submissions/tests.py [hunk 3]
```diff
@@ -58,3 +59,56 @@ def test_multiselect_field_aggregation(self):
...
+        self.assertEqual(expression, 'AA = 1')
+
+        with self.assertRaises(KeyError):
+            expression = build_expression(invalid_check)
+
+    def test_multiple_checks(self):
+        invalid_checks = [
+            {
+                'lvalue': 'AA', 'comparator': '=', 'rvalue': '1',
+                'conjunction': 'and'
```
Comment (takinbo) [inline apollo/submissions/tests.py]: Conjunction should be `&&` not `and`. Did you run tests and have it succeed?

Comment (takinbo) [reply apollo/submissions/tests.py]: I see what you did there. These should be invalid.

File: apollo/submissions/tests.py [hunk 4]
```diff
@@ -58,3 +59,56 @@ def test_multiselect_field_aggregation(self):
...
+            },
+            {
+                'lvalue': 'BH', 'comparator': '=', 'rvalue': 'EJ',
+                'conjunction': 'and'
+            }
+        ]
+        valid_checks = [
+            {
+                'lvalue': 'AA', 'comparator': '=', 'rvalue': '1',
+                'conjunction': 'and'
```
Comment (takinbo) [inline apollo/submissions/tests.py]: Although this should be `&&`

File: apollo/submissions/tests.py [hunk 5]
```diff
@@ -58,3 +59,56 @@ def test_multiselect_field_aggregation(self):
...
+            },
+            {
+                'lvalue': 'BA', 'comparator': '=', 'rvalue': '1',
+                'conjunction': 'and'
+            },
+            {
+                'lvalue': 'BH', 'comparator': '=', 'rvalue': 'EJ',
+                'conjunction': 'or'
```
Comment (takinbo) [inline apollo/submissions/tests.py]: And this `||`

File: pytest.ini [hunk 6]
```diff
@@ -0,0 +1,2 @@
+[pytest]
```
Comment (takinbo) [inline pytest.ini]: Why the explicit need for a `pytest.ini` file?

Comment (dodumosu) [reply pytest.ini]: this just contains a directive to tell `pytest` to look for tests in modules matching the pattern `tests*.py` (which is what we're using here), as opposed to its default of `*_test.py` or `test_*.py`

File: apollo/submissions/qa/query_builder.py [hunk 7]
```diff
@@ -291,3 +294,18 @@ def get_inline_qa_status(submission, condition):
...
     return result, used_tags
+
+
+def build_expression(logical_check):
+    if 'criteria' in logical_check:
+        check_expression = ''
+
+        for index, cond in enumerate(logical_check['criteria']):
+            if index:
+                check_expression += '{conjunction} {lvalue} {comparator} {rvalue} '.format(**cond)
```
Comment (takinbo) [inline apollo/submissions/qa/query_builder.py]: `control_expression`

File: apollo/submissions/qa/query_builder.py [hunk 8]
```diff
@@ -291,3 +294,18 @@ def get_inline_qa_status(submission, condition):
...
+            else:
+                check_expression += '{lvalue} {comparator} {rvalue} '.format(**cond)
```
Comment (takinbo) [inline apollo/submissions/qa/query_builder.py]: `control_expression`

File: apollo/submissions/qa/query_builder.py [hunk 9]
```diff
@@ -291,3 +294,18 @@ def get_inline_qa_status(submission, condition):
...
+    else:
+        check_expression = '{lvalue} {comparator} {rvalue} '.format(**logical_check)
```
Comment (takinbo) [inline apollo/submissions/qa/query_builder.py]: `control_expression`

File: apollo/submissions/qa/query_builder.py [hunk 10]
```diff
@@ -291,3 +294,18 @@ def get_inline_qa_status(submission, condition):
...
+
+    return check_expression.strip()
```
Comment (takinbo) [inline apollo/submissions/qa/query_builder.py]: `control_expression`

File: apollo/submissions/tests.py [hunk 11]
```diff
@@ -58,3 +59,36 @@ def test_multiselect_field_aggregation(self):
 
         result = _multiselect_field_processor(options, column)
         self.assertEqual(result, [4, 4, 5, 4, 4])
+
+
+class ExpressionBuilderTestCase(TestCase):
+    def test_single_check(self):
+        valid_check = {'lvalue': 'AA', 'comparator': '=', 'rvalue': '1'}
```
Comment (takinbo) [inline apollo/submissions/tests.py]: `valid_control`

File: apollo/submissions/tests.py [hunk 12]
```diff
@@ -58,3 +59,36 @@ def test_multiselect_field_aggregation(self):
 
         result = _multiselect_field_processor(options, column)
         self.assertEqual(result, [4, 4, 5, 4, 4])
+
+
+class ExpressionBuilderTestCase(TestCase):
+    def test_single_check(self):
```
Comment (takinbo) [inline apollo/submissions/tests.py]: `test_single_control`

File: apollo/submissions/tests.py [hunk 13]
```diff
@@ -58,3 +59,36 @@ def test_multiselect_field_aggregation(self):
...
+        valid_check = {'lvalue': 'AA', 'comparator': '=', 'rvalue': '1'}
+        invalid_check = {'favourite_fruit': 'apples'}
```
Comment (takinbo) [inline apollo/submissions/tests.py]: `invalid_control`

File: apollo/submissions/tests.py [hunk 14]
```diff
@@ -58,3 +59,36 @@ def test_multiselect_field_aggregation(self):
...
+
+        expression = build_expression(valid_check)
+        self.assertEqual(expression, 'AA = 1')
+
+        with self.assertRaises(KeyError):
+            expression = build_expression(invalid_check)
+
+    def test_multiple_checks(self):
+        valid_checks = {
```
Comment (takinbo) [inline apollo/submissions/tests.py]: `valid_controls`

**Comment by takinbo** [comment]
There are still a couple of issues with this PR:

1. Quality control gets flagged even when there's partial data (e.g. AA = AB is flagged even if there's no data for AB).
2. Quality Assurance warning for a condition is still displayed when all questions are verified.
3. Searching for verified quality control conditions turns up empty even for checklists with verified questions.

**Pull request merged**